For the Wine Quality Data Set, convert all the values in the quality attribute to
0 (bad) if the value is less than or equal to 6 and to 1 (good) otherwise. Normalize
all the other attributes between 0 and 1 by min-max scaling. Mention why we use
min-max scaling.

In [35]:
import pandas as pd
import numpy as np

In [36]:
data = pd.read_csv('WineQT.csv')
print(data.shape)
cols = data.columns.tolist()
print(cols)
data.head()

(1143, 13)
['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar', 'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density', 'pH', 'sulphates', 'alcohol', 'quality', 'Id']


fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  Id  
0      9.4        5   0  
1      9.8        5   1  
2      9.8        5   2  
3      9.8        6   3  
4      9.4        5   4

There are 1143 rows and 13 columns

['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar', 'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density', 'pH', 'sulphates', 'alcohol', 'quality', 'Id']

In [37]:
for i in range(len(data)):
    if (data.loc[i, 'quality'] <= 6):
        data.loc[i, 'quality'] = 0
    else:
        data.loc[i, 'quality'] = 1
data.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  Id  
0      9.4        0   0  
1      9.8        0   1  
2      9.8        0   2  
3      9.8        0   3  
4      9.4        0   4

In [38]:
min_max = data.describe().transpose()[["min", "max"]]
min_max

min         max
fixed acidity         4.60000    15.90000
volatile acidity      0.12000     1.58000
citric acid           0.00000     1.00000
residual sugar        0.90000    15.50000
chlorides             0.01200     0.61100
free sulfur dioxide   1.00000    68.00000
total sulfur dioxide  6.00000   289.00000
density               0.99007     1.00369
pH                    2.74000     4.01000
sulphates             0.33000     2.00000
alcohol               8.40000    14.90000
quality               0.00000     1.00000
Id                    0.00000  1597.00000

Using the above data set, we can get the min and max value for each column as


> min_max.iloc[index,0 if(min) 1 if(max)




In [39]:
min_max.iloc[0,0]

4.6

In Min-Max scaling:

The formula used is:

$$x_{scaled} = \frac{x_i - x_{min}}{x_{max} -x_{min}}$$

In [40]:
#We will consider all the columns except quality and id
cols = cols[0:-2]
for i in range(len(cols)):
  data[cols[i]] = (data[cols[i]] - min_max.iloc[i,0])/(min_max.iloc[i,1] - min_max.iloc[i,0])
data.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0       0.247788          0.397260         0.00        0.068493   0.106845   
1       0.283186          0.520548         0.00        0.116438   0.143573   
2       0.283186          0.438356         0.04        0.095890   0.133556   
3       0.584071          0.109589         0.56        0.068493   0.105175   
4       0.247788          0.397260         0.00        0.068493   0.106845   

   free sulfur dioxide  total sulfur dioxide   density        pH  sulphates  \
0             0.149254              0.098940  0.567548  0.606299   0.137725   
1             0.358209              0.215548  0.494126  0.362205   0.209581   
2             0.208955              0.169611  0.508811  0.409449   0.191617   
3             0.238806              0.190813  0.582232  0.330709   0.149701   
4             0.149254              0.098940  0.567548  0.606299   0.137725   

    alcohol  quality  Id  
0  0.153846        0   0  
1  0.215385        0   1  
2  0.215385        0   2  
3  0.215385        0   3  
4  0.153846        0   4

In [43]:
data.describe().transpose()[["min","max"]]

min     max
fixed acidity         0.0     1.0
volatile acidity      0.0     1.0
citric acid           0.0     1.0
residual sugar        0.0     1.0
chlorides             0.0     1.0
free sulfur dioxide   0.0     1.0
total sulfur dioxide  0.0     1.0
density               0.0     1.0
pH                    0.0     1.0
sulphates             0.0     1.0
alcohol               0.0     1.0
quality               0.0     1.0
Id                    0.0  1597.0

We can see after aplying min_max scaling all the data values are there in the range of 0 and 1, quality is either 0 and 1 based on the condition defined above. and we didnt modified Id as it is not much of our use.